src/pq_flash_index.cpp

```diff
@@ -1077,13 +1092,30 @@ void PQFlashIndex<T, LabelT>::cached_beam_search(const T *query1, const uint64_t
     }
 
     if (beam_width > MAX_N_SECTOR_READS)
+    {
+        context.SetState(State::FailureException);
         throw ANNException("Beamwidth can not be higher than MAX_N_SECTOR_READS", -1, __FUNCSIG__, __FILE__, __LINE__);
+    }
 
+    Timer query_timer, io_timer, cpu_timer;
     ScratchStoreManager<SSDThreadData<T>> manager(this->thread_data);
     auto data = manager.scratch_space();
     IOContext &ctx = data->ctx;
     auto query_scratch = &(data->scratch);
     auto pq_query_scratch = query_scratch->_pq_scratch;
+    if (stats == nullptr)
+    {
+        stats = &context.GetStats();
+    }
+    if (stats != nullptr)
```

Comment:
I think we can get rid of the QueryStats and stick to the one inside IndexSearchContext .